### DISCUSSION TITLE
AVR Port for Micropython

### DISCUSSION THREAD

**Discussion by keenanjohnson** [micropython/micropython#12550] [Other]
Hello community! I was wondering if anyone has worked on a micropython port for an AVR architecture?

I just received a [AVR128DB48 board](https://www.microchip.com/en-us/product/avr128db48#) and I was thinking that it might be great to start working on a micropython port that could run on the board.

**Comment by robert-hh** [comment]
With 128k flash and 16k RAM this could be tricky. More than an absolute minimal system will not fit.

**Comment by keenanjohnson** [comment]
Ah I see and thanks.

Are the recommended memory limits for micorpython documentated anywhere?

- Keenan

**Comment by rkompass** [comment]
The Microbit v1 board also has 16k RAM but 256k Flash, which together appears like a lower limit.
The Micropython compiler was modified to run with such tiny RAM, this is documented somewhere.
With a (Blackpill)  STM32F401 with 64k RAM and 256k Flash standard Micropython runs fine.
With a (Bluepill) STM32F103 with 20k RAM and 64k Flash there only were unsuccessful attempts to port Micropython.

**Comment by robert-hh** [comment]
RAM is the critical limit. With 16 k RAM, reserving 4k for stack and 2k for static data, only 10k remain for the Python heap, which holds the compiled Python code & data.
One more example:
The SAMD21 have 256k flash and 32k RAM. Of the flash 192k are used for code, and 64 k for a small file system. From the RAM, ~22k are available for the heap at 8k stack size.
